### PR TITLE
add support for inline MarkupExtension with named parameters

### LINKED ISSUE #7947: Xaml parsing of Binding with custom converter and expression as converter parameter doesn't work
### Current behavior

I made a custom value converter and a custom markup expression. It basically allows me to bind to a boolean property, and then conditionally return 1 out of 2 values.
See attached project. 
MainPage contains 1 CheckBox and 1 TextBlock. Checking the CheckBox should make the textblock show its text bigger.

Compiling the FontSize expression gives me the following exception:

`MSBUILD : error : Generation failed: System.AggregateException: One or more errors occurred. ---> System.InvalidOperationException: Generation failed for Uno.UI.SourceGenerators.XamlGenerator.XamlCodeGenerator. System.AggregateException: Er zijn één of meer fouten opgetreden. ---> Uno.UI.SourceGenerators.XamlGenerator.XamlParsingException: Expected '}' in the markup extension attribute: '{expressions:ConditionalValuesExpression TrueValue=24' `

### Expected behavior

At runtime, checking the checkbox should make the text in the textblock bigger. 

### How to reproduce it (as minimally and precisely as possible)

Build attached solution. 
I verified, the very same solution works fine on WinUI with MS tooling.

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia, Uno.SourceGenerationTasks

### NuGet package version(s)

- Uno.WinUI*: 4.1.0-dev.542
- 

### Affected platforms

Skia (WPF), Skia (GTK on Linux/macOS/Windows), Build tasks

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Xiaoy312**
GitHub Issue (If applicable): closes #7947

## PR Type

What kind of change does this PR introduce?

- Bugfix
- Feature
## What is the new behavior?

- Binding with MarkupExtension block that contains separator `,` are now parsed correctly
- Custom markup can be used in binding

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information
```
{Binding Converter={StaticResource SomeNotImplementedConverter}, ConverterParameter={ext:MarkupWithArgs Prop1=1, Prop2=2}}
```
^ This was previously split into 3 parameters:
```
Converter={StaticResource SomeNotImplementedConverter}
ConverterParameter={ext:MarkupWithArgs Prop1=1
Prop2=2}
```

### CODE REVIEW

File: src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs [hunk 1]
```diff
@@ -42,6 +42,17 @@ internal class ParsedMarkupExtensionInfo
 		/// - FallbackValue='C,D,E,F'
 		/// </summary>
 		private static Regex BindingMembersRegex = new Regex("[^'\",]+'[^^']+'|[^'\",]+\"[^\"]+\"|[^,]+");
+		private static Regex BalancedMarkupBlockRegex = new Regex(@"
```
Comment (Xiaoy312) [inline src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs:45]: need more test cases against this regex

File: src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs [hunk 2]
```diff
@@ -89,16 +100,24 @@ public static ParsedMarkupExtensionInfo Parse(string raw, IXamlNamespaceResolver
...
-				.Cast<Match>()
-				.Select(m => m.Value.Trim())
+			//var vpairs = BindingMembersRegex.Matches(valueWithoutBinding)
+			//	.Cast<Match>()
+			//	.Select(m => m.Value.Trim())
+			//	.ToList();
+			//if (vpairs.Count == 0)
+			//{
+			//	vpairs.Add(valueWithoutBinding);
+			//}
```
Comment (jeromelaban) [inline src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs:110]: Remove ?

**Review by MartinZikmund — APPROVED** [review]

**Comment by Xiaoy312** [comment]
update20220319: 
- scrapped my initial approach of using balanced regex for the reasons pointed out in comment, since it cannot handle quoted string with comma in it... and, replaced that with a "tokenizer"
- added more test cases to validate all 3 scenarios

File: src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs [hunk 3]
```diff
@@ -63,47 +52,40 @@ public static ParsedMarkupExtensionInfo Parse(string raw, IXamlNamespaceResolver
...
-			var ret = new ParsedMarkupExtensionInfo();
-			int idx = raw.LastIndexOf('}');
+			if (raw.Length >= 2 && raw[1] == '}')
+			{
+				throw Error("Markup extension can not begin with an '{}' escape: '{0}'", raw);
+			}
 
-			if (idx < 0)
+			var ret = new ParsedMarkupExtensionInfo();
+			if (raw[raw.Length - 1] != '}')
```
Comment (xperiandri) [inline src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs:61]:
```suggestion
			if (raw[^1] != '}')
```

File: src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs [hunk 4]
```diff
@@ -63,47 +52,40 @@ public static ParsedMarkupExtensionInfo Parse(string raw, IXamlNamespaceResolver
...
 			}
-				
-			raw = raw.Substring(1, idx - 1);
-			idx = raw.IndexOf(' ');
-			string name = idx < 0 ? raw : raw.Substring(0, idx);
 
-			XamlTypeName xtn;
-			if (!XamlTypeName.TryParse(name, nsResolver, out xtn))
+			var nameSeparatorIndex = raw.IndexOf(' ');
+			var name = nameSeparatorIndex != -1 ? raw.Substring(1, nameSeparatorIndex - 1) : raw.Substring(1, raw.Length - 2);
```
Comment (xperiandri) [inline src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs:69]:
```suggestion
			var name = nameSeparatorIndex != -1 ? raw[1..(nameSeparatorIndex - 1)] : raw[1..^2];
```

File: src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs [hunk 5]
```diff
@@ -63,47 +52,40 @@ public static ParsedMarkupExtensionInfo Parse(string raw, IXamlNamespaceResolver
...
+			if (nameSeparatorIndex < 0)
 				return ret;
 
-			var valueWithoutBinding = raw.Substring(idx + 1, raw.Length - idx - 1);
-
-			var vpairs = BindingMembersRegex.Matches(valueWithoutBinding)
-				.Cast<Match>()
-				.Select(m => m.Value.Trim())
-				.ToList();
+			var valueWithoutBinding = raw.Substring(nameSeparatorIndex + 1, raw.Length - 1 - (nameSeparatorIndex + 1));
```
Comment (xperiandri) [inline src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs:81]: Probably `nameSeparatorIndex + 1` can be named somehow and cached into a variable
```suggestion
			var valueWithoutBinding = raw[(nameSeparatorIndex + 1)..^(1 + (nameSeparatorIndex + 1))];
```

File: src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs [hunk 6]
```diff
@@ -203,5 +185,52 @@ static Exception Error(string format, params object[] args)
...
+				{
+					bracketDepth--;
+					if (bracketDepth > 0)
+					{
+						throw Error("Unexpected '}}' in markup extension: '{0}'", raw);
+					}
+				}
+				else if (c == ',' && bracketDepth == 0)
+				{
+					yield return vargs.Substring(lastSliceIndex + 1, i - lastSliceIndex - 1).Trim();
```
Comment (xperiandri) [inline src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs:223]:
```suggestion
					yield return vargs[(lastSliceIndex + 1)..(i - lastSliceIndex - 1)].Trim();
```

File: src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs [hunk 7]
```diff
@@ -203,5 +185,52 @@ static Exception Error(string format, params object[] args)
...
+					lastSliceIndex = i;
+				}
+			}
+
+			if (bracketDepth > 0)
+			{
+				throw Error("Expected '}}' in markup extension:", raw);
+			}
+
+			yield return vargs.Substring(lastSliceIndex + 1).Trim();
```
Comment (xperiandri) [inline src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs:233]:
```suggestion
			yield return vargs[(lastSliceIndex + 1)..].Trim();
```

**Review by xperiandri — CHANGES_REQUESTED** [review]
I propose using modern C# inderxers syntax for extracting substrings as it simplifies code

**Comment by xperiandri** [comment]
Also, this code can be drastically simplified using [FParsec](http://www.quanttec.com/fparsec/) library

**Comment by jeromelaban** [comment]
There are some issues, but it looks like the unit tests for the System.Xaml assemblies are not run properly. They should be included somehow by this: https://github.com/unoplatform/uno/blob/a9496507a82482ed85d134ba545241ddb2824123/build/ci/.azure-devops-unit-tests.yml#L55-L60

**Comment by Xiaoy312** [comment]
Uno.Xaml.Tests are not being picked up, not because of the `testAssemblyVer2` filter rules
but because it was compiled for net462 and the test runner expected net461 for some reason:

> ##[warning]DiscoveryMessage : Test run detected DLL(s) which were built for different framework and platform versions. Following DLL(s) do not match current settings, which are .NETFramework,Version=v4.6.1 framework and X86 platform.
> ...
> Uno.Xaml.Tests.dll is built for Framework .NETFramework,Version=v4.6.2 and Platform AnyCPU.

I have dropped the TargetFramework for that project, and it compiled locally.
Will see how CI like it~

**Comment by jeromelaban** [comment]
That's great tests got picked up! Thanks :)

**Review by jeromelaban — APPROVED** [review]

**Pull request merged**